pkg/apis/operatorstatus.openshift.io/v1/types.go

```diff
@@ -25,52 +25,81 @@ type OperatorStatus struct {
-	Extension runtime.RawExtension `json:"extension"`
+	Extension runtime.RawExtension `json:"extension,omitempty"`
 }
 
+type ConditionStatus string
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
+// "ConditionFalse" means a resource is not in the condition. "ConditionUnknown" means kubernetes
+// can't decide if a resource is in the condition or not. In the future, we could add other
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
 // OperatorStatusCondition represents the state of the operator's
 // reconciliation functionality.
+// +k8s:deepcopy-gen=true
 type OperatorStatusCondition struct {
 	// Type specifies the state of the operator's reconciliation functionality.
 	Type OperatorStatusConditionType `json:"type"`
 
-	// Message provides any additional information about the current condition.
+	// Status of the condition, one of True, False, Unknown.
+	Status ConditionStatus `json:"status"`
+
+	// LastTransitionTime is the time of the last update to the current status object.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+
+	// reason is the reason for the condition's last transition.  Reasons are CamelCase
+	Reason string `json:"reason,omitempty"`
+
+	// message provides additional information about the current condition.
 	// This is only to be consumed by humans.
-	Message string `json:"message"`
+	Message string `json:"message,omitempty"`
 }
 
 // OperatorStatusConditionType is the state of the operator's reconciliation
 // functionality.
 type OperatorStatusConditionType string
 
 const (
-	// OperatorStatusConditionTypeWaiting indicates that the operator isn't
-	// running its reconciliation functionality. This may be because a
-	// dependency or other prerequisite hasn't been satisfied.
-	OperatorStatusConditionTypeWaiting OperatorStatusConditionType = "Waiting"
-
-	// OperatorStatusConditionTypeWorking indicates that the operator is
-	// actively reconciling its operands.
-	OperatorStatusConditionTypeWorking OperatorStatusConditionType = "Working"
-
-	// OperatorStatusConditionTypeDone indicates that the operator has finished
-	// reconciling its operands and is waiting for changes.
-	OperatorStatusConditionTypeDone OperatorStatusConditionType = "Done"
-
-	// OperatorStatusConditionTypeDegraded indicates that the operator has
-	// encountered an error that is preventing it from working properly.
-	OperatorStatusConditionTypeDegraded OperatorStatusConditionType = "Degraded"
+	// OperatorAvailable indicates that the binary supported by the operator is functional and available in the cluster.
+	OperatorAvailable OperatorStatusConditionType = "Available"
+
+	// OperatorProgressing indicates that the operator is actively binary supported by the operator is functional and available in the cluster.
```

Comment:
Looks like the wrong godoc.

Is progressing supposed to be "I'm reconciling towards the desired state"?  Or reflect the operator status itself?

I think it's acceptable to have different condition types when the definition is not the same.  For instance, Deployment uses Progressing False to represent "failing" because deployment has no perm fail state.

The removal of "Done" means that a consumer doesn't have an unambiguous signal about when all possible changes have been applied.  I think we should have a positive, affirmative condition on "up to date".